databuilder/query_engines/sqlite.py

```diff
@@ -151,9 +282,101 @@ def engine(self):
+
+    This function identifies such placeholder tables, generates an appropriate Common
+    Table Expression which defines what the contents of this table should be, and
+    attaches it to the query.
+    """
+    id_column = query.selected_columns[0]
+    placeholder_table = id_column.table
+    if not getattr(placeholder_table, "is_placeholder", False):
+        return query
+
+    other_tables = [t for t in query.get_final_froms() if t is not placeholder_table]
+    if len(other_tables) > 1:
+        # Select all patient IDs from all tables used in the query
+        id_selects = [
+            sqlalchemy.select(table.c[id_column.name]) for table in other_tables
+        ]
+        # Create a CTE which is the union of all these IDs. (Note UNION rather than
+        # UNION ALL so we don't get duplicates. The tables themselves are necessarily
+        # one-row-per-patient tables and so don't require de-duplicating.)
+        placeholder_definition = sqlalchemy.union(*id_selects).cte(
+            placeholder_table.name
+        )
+        # Include the CTE definition in the query
+        return query.add_cte(placeholder_definition)
+    elif len(other_tables) == 1:
+        # If there's only one table then, rather than define a CTE for it, just replace
+        # references to the placeholder table with references to this table. Using a CTE
+        # wouldn't be wrong — just pointless; and they can act as optimisation fences in
+        # some DBMSs.
+        new_id_column = other_tables[0].c[id_column.name]
+        # SQLAlchemy provides some handy "query surgery" functions, see:
+        # https://docs.sqlalchemy.org/en/14/core/visitors.html#sqlalchemy.sql.visitors.replacement_traverse
+        return replacement_traverse(
+            query, {}, replace=lambda obj: new_id_column if obj is id_column else None
+        )
     else:
-        # Handle literal values (e.g. True) which contain no table references
-        return ()
+        assert False, "No tables found in query"
+
+
+def apply_patient_joins(query):
+    """
+    Find any table references in `query` which aren't yet part of an explicit JOIN and
+    LEFT OUTER JOIN them into the query using the first selected column as the join key
+
+    A core feature of the Query Model/Engine is that we can arbitrarily include data
+    from patient-level tables in a query because, in effect, there is always an implicit
+    join on `patient_id`. This function makes those implicit joins explicit.
+    """
+    # We use the convention that the column to be joined on is always the first selected
+    # column. This avoids having to hardcode, or pass around, the name of the column.
+    join_key = query.selected_columns[0]
+    join_column = join_key.name
+    # Any tables referenced by clauses in the query which aren't yet explicity joined on
+    # will be returned as additional FROM clauses by the `get_final_froms()` method
+    implicit_joins = query.get_final_froms()[1:]
```

Comment:
Why not the first one?

Reply:
I've updated the comment, does this make more sense?
```
# The table referenced by `join_key`, and any tables already explicitly joined with
# it, will be returned as the first value from the `get_final_froms()` method. Any
# remaining tables which aren't yet explicitly joined on will be returned as
# additional values.
```

Reply:
Yes. Is this documented behaviour by SQLAlchemy?

Reply:
I've further updated the comment. I think this code needs all the context it can get.
```
# The table referenced by `join_key`, and any tables already explicitly joined with
# it, will be returned as the first value from the `get_final_froms()` method
# (because `join_key` is the first column). Any remaining tables which aren't yet
# explicitly joined on will be returned as additional clauses in the list. The best
# explanation of SQLAlchemy's behaviour here is probably this:
# https://docs.sqlalchemy.org/en/14/changelog/migration_14.html#change-4737
```